OpenThread

SebastianBoe opened the pull request:
This PR contains OpenThread support

`cmake -Boe -DBOARD=nrf52840_pca10056 -DCONF_FILE=prj_nrf52840_ot.conf -Hsamples/net/echo_server && make -C oe`

This will download OpenThread sources and build an application that uses them.

One can also point to a local OpenThread repo by doing:

cmake -DEXTERNAL_PROJECT_PATH_OPENTHREAD=path/to/openthread

(The intention is to change this CMake variable into a Kconfig eventually)

PS: More extensive testing has been done earlier, but those results are no longer valid.

galak commented:
We've gone around and around on this, but what's the feeling of having non Apache licesnsed code in subsys?

galak commented on drivers/ieee802154/Kconfig:
```diff
@@ -51,4 +51,8 @@ config IEEE802154_UPIPE_DRV_NAME
+config NETWORKING_WITH_15_4
+	bool
+	default n
+
```
we have config IEEE802154, so why do we need this?

tbursztyka replied:
see patch "include: net: net_pkt: Use NETWORKING_WITH_15_4 "

galak commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -441,7 +441,7 @@ NET_DEVICE_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
 		CONFIG_IEEE802154_NRF5_INIT_PRIO,
 		&nrf5_radio_api, OPENTHREAD_L2,
-		NET_L2_GET_CTX_TYPE(OPENTHREAD_L2), 125);
+		NET_L2_GET_CTX_TYPE(OPENTHREAD_L2), 1280);
```
why is this change in the 'subsys: net: l2: Add OpenThread L2' commit?

KaSroka replied:
Not sure how it got here, I will clean commits after CMake porting is finished.

galak reviewed (changes requested):
Made a few comments, but the commit messages need significant cleaning up to explain what is going on here.  How is this integration occurring - how are we glue the zephyr network stack & openthread together?

SebastianBoe commented:
> How is this integration occurring - how are we glue the zephyr network stack & openthread together?

Not sure if I am answering your question. But the CMake build scripts for the networking stack will invoke OpenThread's build system in a similar manner to how the external_lib sample works.

nashif commented:
> We've gone around and around on this, but what's the feeling of having non Apache licesnsed code in subsys?

which code is non Apache?

galak commented:
> which code is non Apache?

I guess I saw it on a Makefile in the ext/ import and assumed everything else was as well.

galak commented:
> How is this integration occurring - how are we glue the zephyr network stack & openthread together?
> Not sure if I am answering your question. But the CMake build scripts for the networking stack will invoke OpenThread's build system in a similar manner to how the external_lib sample works.

No, I'm asking how packets & drivers interfaces are working, not at a build level (although that should also be documented).

KaSroka commented:
> I guess I saw it on a Makefile in the ext/ import and assumed everything else was as well.
Everything from ext/lib/openthread will be moved to subsys/net/lib/openthread so it will have Zephyr license

KaSroka commented:
> No, I'm asking how packets & drivers interfaces are working, not at a build level (although that should also be documented).

Main part of this glue layer is in subsys/net/ip/l2/openthread. Generally OpenThread is implemented as L2 layer in Zephyr.

tbursztyka commented on subsys/net/ip/l2/openthread/openthread.c:
```diff
@@ -0,0 +1,348 @@
+		while (otTaskletsArePending(ot_context->instance)) {
+			otTaskletsProcess(ot_context->instance);
+		}
+		PlatformProcessDrivers(ot_context->instance);
```
try to break a line after a '}' unless it's another } etc...

tbursztyka commented on subsys/net/ip/l2/openthread/openthread.c:
```diff
@@ -0,0 +1,348 @@
+}
+
+static enum net_verdict openthread_recv(struct net_if *iface,
+	struct net_pkt *pkt)
```
so as said above, parameter indentation below last '(' + 1 space is also applied to function declaration

tbursztyka commented on subsys/net/ip/l2/openthread/openthread.c:
```diff
@@ -0,0 +1,348 @@
+		goto exit;
+	}
+
+	struct net_buf *frag;
```
variable declaration is always at the beginning of a code block, so move this above.

tbursztyka commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,293 @@
+
+#if defined(CONFIG_OPENTHREAD_L2_DEBUG_DUMP_15_4) || \
+	defined(CONFIG_OPENTHREAD_L2_DEBUG_DUMP_IPV6)
+void dump_pkt(const char *str, struct net_pkt *pkt)
```
Remove that function and use net_hexdump_frags() from net_private.h instead

tbursztyka commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,293 @@
+	}
+
+	i_idx++;
+
```
you can remove that empty line, as the below condition is testing the previous assignment on i_idx.

tbursztyka commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,293 @@
+	}
+
+	context->pkt_list_out_idx++;
+
```
same here

tbursztyka commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,293 @@
+			continue;
+		}
+
+		const otNetifAddress *address;
```
move the 2 lines a the beginning of for {} block

tbursztyka commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,293 @@
+			continue;
+		}
+
+		const otNetifMulticastAddress *maddress;
```
same

tbursztyka commented on drivers/ieee802154/Kconfig.nrf5:
```diff
@@ -16,7 +16,6 @@ config IEEE802154_NRF5_DISABLED
 
 config IEEE802154_NRF5
 	bool "Regular mode"
-	select NET_L2_IEEE802154
```
instead you should be able to do:
select NET_L2_IEEE802154 if not <ot l2 config>

Afaik. or is it ! instead of not? can't recall.

tbursztyka commented on ext/Kconfig:
```diff
@@ -12,6 +12,8 @@ source "ext/hal/Kconfig"
 
 source "ext/lib/crypto/Kconfig"
 
+source "ext/lib/openthread/Kconfig"
```
merge this patch with "ext: lib: Add OpenThread to ext/lib"

tbursztyka commented on subsys/net/ip/l2/Kconfig:
```diff
@@ -132,4 +132,6 @@ config NET_DEBUG_ARP
 	help
```
Merge this patch with "subsys: net: l2: Add OpenThread L2 "

tbursztyka reviewed (changes requested):
Check your editor, but it should be able to apply Linux code style for you, so automatically indenting parameters relevantly.

nashif commented:
run the code through:

`uncrustify -c ${ZEPHYR_BASE}/scripts/uncrustify.cfg --no-backup --if-changed *.c`

This will cleanup most of the style issues

KaSroka commented:
I will start addressing those request changes as soon as @SebastianBoe finishes moving to CMake

carlescufi commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -97,7 +97,8 @@ static void nrf5_rx_thread(void *arg1, void *arg2, void *arg3)
 		 */
-#if defined(CONFIG_IEEE802154_NRF5_RAW)
+#if defined(CONFIG_IEEE802154_NRF5_RAW) || \
+	defined(CONFIG_NET_L2_OPENTHREAD)
```
Do we need to update the rest of 15.4 drivers with something similar to this?

tbursztyka replied:
Let's see first how the whole behave with nrf5.

I am not entirely sure I'll want to keep this that way in the future (actually, I am not found of CONFIG_IEEE802154_*_RAW stuff either).

carlescufi commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -435,6 +436,17 @@ DEVICE_AND_API_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
 		    POST_KERNEL, CONFIG_IEEE802154_NRF5_INIT_PRIO,
 		    &nrf5_radio_api);
+#elif defined(CONFIG_NET_L2_OPENTHREAD)
+NET_DEVICE_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
```
Same here, is a change like this required for the rest of 15.4 drivers?

galak commented:
> Let's see first how the whole behave with nrf5.
> 
> I am not entirely sure I'll want to keep this that way in the future (actually, I am not found of CONFIG_IEEE802154_*_RAW stuff either).

If we keep CONFIG_IEEE802154_*_RAW, seems like that should become CONFIG_IEEE802154_RAW, and have the drivers all support it.  Seems a bit of a mess what we have with CONFIG_IEEE802154_{CC2520, MCR20A, NRF5}_RAW

carlescufi commented:
This PR will now be used to add an external framework for libraries with CMake, and Thread-related comments will be addressed later once it builds with CMake

jukkar commented on subsys/net/ip/l2/openthread/Kconfig:
```diff
@@ -0,0 +1,137 @@
+menuconfig NET_L2_OPENTHREAD
+	bool
+	prompt "OpenThread L2"
+	depends on FLASH
```
Using "depends on" is better than "select" but isn't it a bit cumbersome here. I mean that if FLASH is not selected in some other part of the system, then this option will not be seen in menuconfig for example. I am ok with this thou, just a note.

jukkar commented on subsys/net/lib/openthread/zephyr-mbedtls-config.h:
```diff
@@ -0,0 +1,6 @@
+
+#define MBEDTLS_PLATFORM_EXIT_ALT
+
+#endif /* ZEPHYR_MBEDTLS_CONFIG_H_ */
```
The file should end with \n

SebastianBoe commented:
As agreed upon in the Thread meeting, a branch has been force pushed to this PR branch to enable collaboration.

See PR text for details.

SebastianBoe commented:
There has been a bad merge at some point, so that will be the first item that I would fix here.

This branch has references to CONFIG_NETWORKING_WITH_15_4, but that option
should not exist anymore. Replacing it with CONFIG_NET_L2_IEEE802154 led
to a tricky Kconfig issue I was not able to figure out.

ekarlso commented:
Possible to get a rebase of this? I am keen on testing on KW41Z and a mightygecko.

KaSroka committed:
include: net: net_pkt: Use IEEE802154 instead of NET_L2_IEEE802154

Use CONFIG_IEEE802154 instead of CONFIG_NET_L2_IEEE802154
and CONFIG_IEEE802154_RAW_MODE when adding RSSI and LQI
members to net_pkt. CONFIG_IEEE802154 is selected when
at least one 15.4 device is selected (L2 or RAW).

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka committed:
subsys: net: lib: Add OpenThread platform

OpenThread requires platform definition with standarized API
so we have to add wrappers to make it compatible with Zephyr.
OpenThread is based on autoconf, this requires
more specific CMakeLists.txt which allows to clone specific
commit or point to local copy of openthread.

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka committed:
include: net: net_l2: Add OpenThread L2

Add OpenThread L2 to L2 list.

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka committed:
drivers: 15.4: nrf5: Add OpenThread support to 802.15.4 driver

Initially add support for OpenThread only for nRF 802.15.4 radio driver.

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka committed:
subsys: net: ip: l2: Add OpenThread L2

Add OpenThread to Zephyrs net stack as data link layer.
OpenThread requires to call process function when an event occurs.
This process function is called from cooperative thread.

Packet conversion and dispaching is implemented in openthread.c
as well as addresses forwarding.

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka committed:
samples: net: Add config and enable CI for OpenThread on nRF52840

Add samples config for nRF52840 with OpenThread
for echo_client and echo_server.

Add OpenThread to CI for echo_client and echo_server on nRF52840

Signed-off-by: Kamil Sroka <kamil.sroka@nordicsemi.no>

KaSroka commented:
I've addressed comments so far, cleaned up commit history and fixed any CI errors

KaSroka commented:
This was tested on Nordic dev kit PCA10056 (version 0.9.0) with nRF52840. Samples echo_client and echo_server was used.

To pass Thread certification some improvements will be needed in the future. But right now it can coexist with bare metal OpenThread network.

KaSroka commented:
@ekarlso I've rebased it so you can test it, you will have to adapt 15.4 driver for mightygecko and KW41Z to use it as now only nRF52840 is supported.

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -124,7 +124,7 @@ struct net_pkt {
 #endif /* CONFIG_NET_IPV6 */
 
-#if defined(CONFIG_NET_L2_IEEE802154) || defined(CONFIG_IEEE802154_RAW_MODE)
+#if defined(CONFIG_IEEE802154) || defined(CONFIG_IEEE802154_RAW_MODE)
```
no CONFIG_IEEE802154 is not "more generic". CONFIG_IEEE802154 means one or more 15.4 drivers are built-in. And it does not say about whether 15.4's L2 is being used or if it's the RAW mode.

Change your commit message to say that CONFIG_IEEE802154 matches both case: either L2 or RAW mode, because it means one (or more) 15.4 device driver is being built-in.

And on your patch, remove the check on RAW_MODE

tbursztyka commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -90,7 +94,8 @@ static void nrf5_rx_thread(void *arg1, void *arg2, void *arg3)
 		 */
-		if (IS_ENABLED(CONFIG_IEEE802154_RAW_MODE)) {
+		if (IS_ENABLED(CONFIG_IEEE802154_RAW_MODE) ||
+		    IS_ENABLED(CONFIG_NET_L2_OPENTHREAD)) {
```
Just make the necessary change so that NET_L2_OPENTHREAD selects RAW MODE, so you won't have to do that check here.

KaSroka replied:
Is it a good idea to select raw mode? As raw mode selects NET_RAW_MODE

tbursztyka replied:
good point, forgot about that. Keep your modification then.

galak replied:
We should figure a solution that doesn't require modifying all the drivers to enable openthread.

tbursztyka commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -428,6 +433,17 @@ DEVICE_AND_API_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
+NET_STACK_INFO_ADDR(RX, nrf5_154_radio,
+		    CONFIG_IEEE802154_NRF5_RX_STACK_SIZE,
+		    CONFIG_IEEE802154_NRF5_RX_STACK_SIZE,
+		    nrf5_data.rx_stack, 0);
```
too many lines for just 3 parameters that change.

You should be able to keep only one NET_DEVICE_INIT and one NET_STACK_INFO_ADDR calls, the 3 parameters that changes should be set via in-betweed macros and that should do it.

KaSroka replied:
I've tried this:
```c
#if defined(CONFIG_NET_L2_IEEE802154)
#define L2 IEEE802154_L2
#define MTU 125
#elif defined(CONFIG_NET_L2_OPENTHREAD)
#define L2 OPENTHREAD_L2
#define MTU 1280
#endif

#if defined(CONFIG_NET_L2_IEEE802154) || defined(CONFIG_NET_L2_OPENTHREAD)
NET_DEVICE_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
		nrf5_init, &nrf5_data, &nrf5_radio_cfg,
		CONFIG_IEEE802154_NRF5_INIT_PRIO,
		&nrf5_radio_api, L2,
		NET_L2_GET_CTX_TYPE(L2), MTU);

NET_STACK_INFO_ADDR(RX, nrf5_154_radio,
		    CONFIG_IEEE802154_NRF5_RX_STACK_SIZE,
		    CONFIG_IEEE802154_NRF5_RX_STACK_SIZE,
		    nrf5_data.rx_stack, 0);
#else
DEVICE_AND_API_INIT(nrf5_154_radio, CONFIG_IEEE802154_NRF5_DRV_NAME,
		    nrf5_init, &nrf5_data, &nrf5_radio_cfg,
		    POST_KERNEL, CONFIG_IEEE802154_NRF5_INIT_PRIO,
		    &nrf5_radio_api);
#endif
```
But there is a problem with `NET_L2_GET_CTX_TYPE()` expansion as it wil expand to L2_CTX_TYPE and I cannot see a way to expand it to correct name.

galak commented:
Can we get this ported to a second 15.4 driver platform (kw41z, CC1200/CC2520, etc.)

galak commented on subsys/net/ip/l2/openthread/Kconfig:
```diff
@@ -0,0 +1,195 @@
+	help
+	  Enable jam detection in OpenThread stack
+
+config OT_PLAT_FLASH_DEVICE_NAME
```
Can we get this from device tree?

KaSroka replied:
How to get this info from DT? I didn't liked it as well but I don't know other method.

nvlsianpu replied:
So far DT doesn't describe driver names for any flash controller and nordic RADIO driver.  We would add such description in another patch. Can this workaround be accepted as is?

galak commented on subsys/net/ip/l2/openthread/Kconfig:
```diff
@@ -0,0 +1,195 @@
+	prompt "Flash device name"
+	default "NRF5_FLASH"
+
+config OT_PLAT_RADIO_DEVICE_NAME
```
Can we get this from dt?

galak reviewed (changes requested):
We should use DT to deal with FLASH & DRIVER info

galak reviewed (changes requested):
Can we pull out the NRF5 change to support RAW mode out of this PR.

jukkar commented on subsys/net/ip/l2/openthread/openthread.c:
```diff
@@ -0,0 +1,350 @@
+#endif
+
+exit:
+
```
This extra line can be removed

jukkar commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,247 @@
+	otIp6AddUnicastAddress(context->instance, &addr);
+
+#if CONFIG_OPENTHREAD_L2_LOG_LEVEL == SYS_LOG_LEVEL_DEBUG
+	char buf[NET_IPV6_ADDR_LEN];
```
Can you add {} so that buf is defined in the start of the block? I am not sure if any compiler will complain about this thou.

jukkar commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,247 @@
+	struct otIp6Address addr;
+	int i;
+
+	/* save the last added IP address for this interface */
```
This works atm, but at some point the IP address handling will be changed in net_if (in order to save some memory) so this needs to be changed at that time (currently this comment is just a note to myself).

jukkar commented on subsys/net/ip/l2/openthread/openthread_utils.c:
```diff
@@ -0,0 +1,247 @@
+	otIp6SubscribeMulticastAddress(context->instance, &addr);
+
+#if CONFIG_OPENTHREAD_L2_LOG_LEVEL == SYS_LOG_LEVEL_DEBUG
+	char buf[NET_IPV6_ADDR_LEN];
```
Add {} block here.

jukkar reviewed (commented):
Some minor nitpicks from me.

nashif commented:
> Can we get this ported to a second 15.4 driver platform (kw41z, CC1200/CC2520, etc.)

can this be done in another PR by those who have the hardware and understand it and with help from the people doing this PR if needed? I do not see how this should block this PR

galak commented:
> can this be done in another PR by those who have the hardware and understand it and with help from the people doing this PR if needed? I do not see how this should block this PR

I think it should, it shows that the code is actually able to work on multiple 15.4 drivers w/o modification.

KaSroka commented:
@galak @nashif Bear in mind comment from subsys/net/lib/openthread/platform/radio.c:71 
```
"Only radios with automatic ack handling are currently supported"
```
It's the only type of radio I can test on so I cannot provide code for radios without auto ack capability (IEEE802154_HW_TX_RX_ACK)

jfischer-phytec-iot commented:
@KaSroka I will test it with the MCR20A driver.

nashif commented:
@jfischer-phytec-iot 

> @KaSroka I will test it with the MCR20A driver.

thanks

MaureenHelm commented:
Are the build instructions at the top of the PR up-to-date? I'm having trouble building, getting the following error:
```
[ 18%] Performing configure step for 'ot'
/bin/sh: 1: ./configure: not found
zephyr/subsys/net/lib/openthread/CMakeFiles/ot.dir/build.make:107: recipe for target 'zephyr/ext_proj/Stamp/ot/ot-configure' failed
```

nashif commented:
> Are the build instructions at the top of the PR up-to-date?

works for me, in samples/net/echo_server
```
mkdir build; cd build
cmake -DCONF_FILE=prj_nrf52840_ot.conf -DBOARD=nrf52840_pca10056 ..
make
```

nashif reviewed (changes requested):
please add this configuration to sample.yaml to make it build in CI, for example add:

```
  test_openthread:
    extra_args: CONF_FILE="/prj_nrf52840_ot.conf"
    platform_whitelist: nrf52840_pca10056
```

KaSroka commented:
I've addressed comments so far except comments from @galak as I don't know how to implement them - how can I get flash and radio device from device tree?

carlescufi commented:
@KaSroka not sure about the radio device, but the flash can be obtained via the generated header file, see here: `<build folder/zephyr/include/generated/generated_dts_board.h`

jfischer-phytec-iot commented:
@KaSroka @tbursztyka I tried to build it for FRDM-K64F and MCR20A. For Kinetis SoC, this PR depends on #4633, what is not so dramatic, hacked it quickly together. More problematic is the dependency on `ieee802154_is_ar_flag_set` inside `ieee802154_frame.c,` since the mcr20a_tx has to check whether AR was set or not. I do not have a quick solution for that now.

tbursztyka commented:
@jfischer-phytec-iot It should not be too much of a problem. Some Kconfig option needed in l2 side, so it should be possible to build the framing part in 15.4 l2. (linking will drop the unnecessary code). Anyway, let's get this OT support for nrf first, we will address other chip support in other PR.

jukkar reviewed (approved):
+1 from me.

jfischer-phytec-iot commented on subsys/net/lib/openthread/platform/flash.c:
```diff
@@ -0,0 +1,108 @@
+	pages_count = flash_get_page_count(flash_dev);
+
+	if (flash_get_page_info_by_idx(flash_dev,
+		pages_count - CONFIG_OT_PLAT_FLASH_PAGES_COUNT - 1, &info)) {
```
@KaSroka (for my understanding) what is the valid value range of CONFIG_OT_PLAT_FLASH_PAGES_COUNT? Note: the return value will not be checked.

KaSroka replied:
No less than 2 pages but also you cannot overwrite data in flash. Current implementation uses CONFIG_OT_PLAT_FLASH_PAGES_COUNT pages located at the end of FLASH so if this count is too big it will collide with program data.

jfischer-phytec-iot commented on subsys/net/lib/openthread/platform/flash.c:
```diff
@@ -0,0 +1,108 @@
+	return OT_ERROR_NONE;
+}
+
+u32_t utilsFlashWrite(u32_t aAddress, uint8_t *aData, u32_t aSize)
```
@KaSroka It does not take into account the `write_block_size` and so it will not work on all platforms, e.g. not on NXP K64F. aSize is probably always multiple of 4, is it configurable? I did not find anything suitable in the OT examples.

KaSroka replied:
It is at least multiple of 4 as far as I can see in OT code but it's not configurable.

carlescufi commented:
@tbursztyka are there still outstanding changes required on your side?

jfischer-phytec-iot commented:
It has impressive RAM consumption (FRDM-K64F + MCR20A, similar for pca10056), is it something to expect from OT for a RED, or can it be optimized?
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      200500 B         1 MB     19.12%
            SRAM:       67596 B       192 KB     34.38%
        IDT_LIST:         148 B         2 KB      7.23%
```

KaSroka commented:
@jfischer-phytec-iot Yes this ram usage is expected

KaSroka commented:
@galak Can you take a look at this PR?

galak commented on subsys/net/ip/l2/openthread/Kconfig:
```diff
@@ -0,0 +1,195 @@
+	help
+	  Enable jam detection in OpenThread stack
+
+config OT_PLAT_FLASH_DEVICE_NAME
```
This should now be fixed / merged for Nordic and several other platforms.  We should be able to just use FLASH_DEV_NAME

KaSroka commented:
Rebased and addressed @galak comment. Similar change will be required when #5791 will be merged to master.

dleach02 commented:
To follow-up on a question Maureen asked, what additional build tools need to be installed to support this branch? I'm getting a similar error even using the command line Nashif suggested.

nashif commented:
> To follow-up on a question Maureen asked, what additional build tools need to be installed to support this branch? I'm getting a similar error even using the command line Nashif suggested.

Maybe you need autotools? (autoconf, automake, etc.)

dleach02 commented:
The missing package is 'libtool' which installs libtoolize. I'm guessing we need to add this package to the installation and setup instructions for building Zephyr as the current instructions do not include this package.

carlescufi commented:
@dleach02 I don't think we can document the required tools for all external projects that will integrate with Zephyr. Instead what I would suggest is to link to OpenThread's documentation, although I can't find build instructions on their GitHub or webpage.
